### PR TITLE
Fix mod_subst wrt universe polymorphism

### PR DESCRIPTION
**Pull request by SkySkimmer**
@maximedenes this is how I would do #8971 (and as such is incompatible with it).

Based on #8972.

### CODE REVIEW

File: kernel/vars.mli
Comment (maximedenes) [inline kernel/vars.mli]: unabstract -> apply ?

**Review by maximedenes — COMMENTED** [review]
I like it, but it is disappointing that we have to expose `univ_abstracted`. Is there any hope we can fix the problems (in clients), so that we can expose a safer API?

**Comment by SkySkimmer** [comment]
>I like it, but it is disappointing that we have to expose univ_abstracted. Is there any hope we can fix the problems (in clients), so that we can expose a safer API?

I don't know, there's no point forcing people to provide an instance and substitute when we're just going to look for the head globref or whatever, no? And I'm not unhappy with the univ_abstracted, someday I'll try to add some in const_body.

File: kernel/vars.mli
Comment (SkySkimmer) [reply kernel/vars.mli]: Only if you insist.

**Comment by maximedenes** [comment]
> I don't know, there's no point forcing people to provide an instance and substitute when we're just going to look for the head globref or whatever, no?

But then we should provide an API that inlines only to globrefs, I guess... We wouldn't go through `constr`.

File: kernel/vars.mli
Comment (mattam82) [reply kernel/vars.mli]: `instantiate` would fit even better with the rest of the naming scheme if I may.

**Review by mattam82 — APPROVED** [review]
The code looks good to me, I defer to @maximedenes about the cleanup of the mod_subst API otherwise.

File: kernel/vars.mli
Comment (maximedenes) [reply kernel/vars.mli]: Indeed!

**Comment by SkySkimmer** [comment]
For the [eq -> eq t] inlining that won't work

Gaëtan Gilbert

**Comment by maximedenes** [comment]
> For the [eq -> eq t] inlining that won't work

Why not? The head globref will still be the new `eq`, right?

**Comment by SkySkimmer** [comment]
But it's not a globref -> globref inlining, it's globref -> constr.

**Comment by maximedenes** [comment]
> But it's not a globref -> globref inlining, it's globref -> constr.

But precisely, if the client is concerned only about the head, like you were writing, we could encapsulate that as a globref -> globref inlining.

**Comment by SkySkimmer** [comment]
That makes no sense to me TBH

Gaëtan Gilbert

File: tactics/hints.ml
Comment (herbelin) [inline tactics/hints.ml]: An alternative could be to remove the hint if no head constant appears anymore.

File: kernel/univ.ml
Comment (herbelin) [inline kernel/univ.ml]: You could also use the conventional name of "closure". 

Also, at first, I did not get that the projection `univ_abstracted` meant the body. So, maybe `univ_closure_body` and `univ_closure_binders`?

File: kernel/univ.mli
Comment (herbelin) [inline kernel/univ.mli]: The purists would say "de Bruijn"

File: kernel/univ.ml
Comment (herbelin) [inline kernel/univ.ml]: By the way, why a list of names rather than a `AUContext.t`? Even if you explain explicitly later that the bindings are à la de Bruijn, it is a bit misleading to see names, as if the universes were bound by names.

**Review by herbelin — COMMENTED** [review]
Very nice.

File: pretyping/detyping.ml
Comment (SkySkimmer) [reply pretyping/detyping.ml]: However u is not an instance (it's a glob_instance ie a list of level names) so it's not easy to do.

Comment (herbelin) [reply pretyping/detyping.ml]: This code is basically for substituting the `glob_constr` occurring in LtacX code. At first view, in such code, the instance can only be made of global universes. Would using `Pretyping.interp_instance` be ok?

More upstream, it seems that Ltac code should detect universe errors earlier. For instance, the following leads to a wrong-instance-length runtime error:
```
Ltac f x := exact (id@{Type} x). 
```
and this would also break at the time of an `Inline` substitution of functor arguments.

File: kernel/univ.ml
Comment (SkySkimmer) [reply kernel/univ.ml]: Wouldn't a closure contain values for the universes rather than binder names?

File: kernel/univ.mli
Comment (SkySkimmer) [inline kernel/univ.mli]: Changed to use just an `int` as we have no use for more, if that changes we can put whatever we need in the type later.

File: kernel/univ.ml
Comment (ppedrot) [reply kernel/univ.ml]: I support reusing the `AUContext.t` type. It is more informative, and that information might end up being needed, or allowing to thread data in a uniform way.

Comment (herbelin) [reply kernel/univ.ml]: > Wouldn't a closure contain values for the universes rather than binder names?

You're right. My confusion.

I still feel that the current terminology can be made more appealing though. So, maybe `univ_abstracted_body` and `univ_abstracted_binders`. Or `abstraction` in place of `abstracted`. Of course, that's pretty subjective, so maybe other opinions would help to move from subjective to more objective.

**Comment by SkySkimmer** [comment]
Now using an `AUContext.t` and names `univ_abstracted_value` `univ_abstracted_binder`.

File: interp/notation_ops.ml
Comment (ppedrot) [inline interp/notation_ops.ml]: This is wrong, but if you hit this bug you probably deserve it.

File: kernel/mod_subst.ml
Comment (ppedrot) [inline kernel/mod_subst.ml]: You should probably preserve this comment.

File: kernel/mod_subst.mli
Comment (ppedrot) [inline kernel/mod_subst.mli]: Remove this function.

File: kernel/vars.mli
Comment (ppedrot) [inline kernel/vars.mli]: This is a dangerous function because it drops the universe constraints, and nothing indicates this.

File: tactics/hints.ml
Comment (ppedrot) [reply tactics/hints.ml]: This is indeed a good idea.

**Review by ppedrot — CHANGES_REQUESTED** [review]
Not all the comments are mandatory, but it's probably better to polish this.

File: kernel/vars.mli
Comment (SkySkimmer) [reply kernel/vars.mli]: >This is a dangerous function because it drops the universe constraints, and nothing indicates this.

Should we just go back to carrying an `int` then.

File: kernel/mod_subst.mli
Comment (SkySkimmer) [reply kernel/mod_subst.mli]: Changed.

File: kernel/mod_subst.ml
Comment (SkySkimmer) [reply kernel/mod_subst.ml]: Changed.

File: kernel/vars.mli
Comment (ppedrot) [reply kernel/vars.mli]: A comment is enough!

**Comment by SkySkimmer** [comment]
Have your comment then.

Gaëtan Gilbert

**Comment by SkySkimmer** [comment]
@ppedrot re ping

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
I am not the one assigned to this PR, and my comments were not that mandatory to be taken into account, but here is your approval.

**Review by maximedenes — APPROVED** [review]

**Comment by maximedenes** [comment]
Waiting for a green light from @herbelin.

**Comment by herbelin** [comment]
I find it clearer now. Thanks. Ok to me

**Pull request merged**
